File: OptibenchOptimizer/Implementations/Monitor.cs
using System.Net.Http.Headers;
using System.Text;
using Dtos;
using interfaces;
using Newtonsoft.Json;

namespace Implementations
{
    class Monitor : IMonitor
    {
        private readonly HttpClient client = new HttpClient
        {
           Timeout = TimeSpan.FromSeconds(3) // ako ne dobije odgovor u roku od 3s, http zahtjev ce se prekinuti
        };

        public Monitor(string uri) 
        {
            client.BaseAddress = new Uri(uri); //za "monitor",
            client.DefaultRequestHeaders.Accept.Clear();
            client.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
        }
        public async Task Save(OptimizationResultDto result)  //nema povr vrijednost za sada
        {
            if(double.IsNaN(result.Y))
            {
                Console.WriteLine($"Failed to save the result to the database."); 
                return;
            }
                  

            string path = $"result";
            var resultJson = JsonConvert.SerializeObject(result); // Serijalizujem objekat u json string
            var httpContent = new StringContent(resultJson, Encoding.UTF8, "application/json"); // StringContent moze da se koristi sa klijentom

            var httpResponse = await client.PostAsync(path, httpContent);
            if (!httpResponse.IsSuccessStatusCode)
            {            
                 Console.WriteLine($"Failed to save the result to the database.");  
            }
        }

        
    }
}

File: PythonOptimizer/Implementations/Monitor.py
import math
from urllib.parse import urljoin
import requests
import json
from Dtos import OptimizationResultDto
from Abstractions.IMonitor import IMonitor
from Implementations.RemoteProblem import RemoteProblem

class Monitor(IMonitor):
    def __init__(self, uri: str) -> None:
        self.uri = uri
        self.client = requests.Session()
        self.client.headers.clear()
        self.client.headers.update({'Accept': 'application/json'})


    async def save(self, result: OptimizationResultDto, problem: RemoteProblem) -> None:
        if math.isnan(result.Y):
            print("Failed to save the result to the database.")
            return
        problem_info = json.loads(result.ProblemInfo)
        problem_name = problem_info["ProblemName"]
        print(problem_info)
        print(problem_name)
        result.ExactSolution = await problem.get_exact_solution(problem_name)
        print(result.ExactSolution)

        path = "result"
        full_url = urljoin(self.uri, path)
        result_json = json.dumps(result.__dict__)
        headers = {'Content-Type': 'application/json'}

        print(result_json)
        
        try:
            http_response =  self.client.post(full_url, data=result_json, headers=headers, timeout=3)
            http_response.raise_for_status()
        except requests.exceptions.RequestException as e:
            print(f"Failed to save the result to the database. Error: {e}")
